modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java

```diff
@@ -2055,58 +2061,165 @@ public void visitRegex(ERegex userRegexNode, SemanticScope semanticScope) {
+                    javaFlags |= Pattern.UNICODE_CASE;
                     break;
                 case 'x':
-                    constant |= Pattern.COMMENTS;
+                    javaFlags |= Pattern.COMMENTS;
                     break;
                 default:
                     throw new IllegalArgumentException("invalid regular expression: unknown flag [" + flag + "]");
             }
         }
+        return javaFlags;
+    }
 
+    private Grok compileGrok(ERegex regex, SemanticScope semanticScope) {
+        if (false == "".equals(regex.getFlags())) {
+            throw regex.createError(
+                new IllegalArgumentException("Grok doesn't support flags but got [" + regex.getFlags() + "]")
+            );
+        }
         try {
-            Pattern.compile(pattern, constant);
-        } catch (PatternSyntaxException pse) {
-            throw new Location(location.getSourceName(), location.getOffset() + 1 + pse.getIndex()).createError(
-                    new IllegalArgumentException("invalid regular expression: " +
-                            "could not compile regex constant [" + pattern + "] with flags [" + flags + "]", pse));
+            
+            Map<String, String> patternBank = semanticScope.getScriptScope().getCompilerSettings().getGrokPatternBank();
+            MatcherWatchdog watchdog = semanticScope.getScriptScope().getGrokWatchdog();
+
+            /*
+             * Shift from the "no permissions" compilation context to one that
+             * has permission to read stuff from the classpath so that jcodings,
+             * a dependency of grok, can read what it needs from the classpath. 
+             */
+            return AccessController.doPrivileged(new PrivilegedAction<Grok>() {
+                @Override
+                public Grok run() {
+                    // Try to collect warnings up front and refuse to compile the expression if there are any
+                    List<String> warnings = new ArrayList<>();
+                    new Grok(patternBank, regex.getPattern(), watchdog, warnings::add).match("___nomatch___");
+                    if (false == warnings.isEmpty()) {
+                        throw regex.createError(
+                            new IllegalArgumentException(
+                                "Compiled grok pattern constant [" + regex.getPattern() + "] but emitted warnings: " + warnings
+                            )
+                        );
+                    }
+
+                    return new Grok(
+                        patternBank,
+                        regex.getPattern(),
+                        watchdog,
+                        w -> {
+                            throw new IllegalArgumentException("grok [" + regex.getPattern() + "] emitted a warning: " + w);
+                        }
+                    );
+                }
+            });
+        } catch (IllegalArgumentException e) {
+            Location errorLocation = new Location(
```

Comment:
Side point: a `Location.offset` method that returns another location offset by some characters would come in handy here.